### PR TITLE
New linear solver factory

### PR DESCRIPTION
**Pull request by loumalouomega**
I need help to migrate and improve the new_linear_solver_factory

### CODE REVIEW

**Comment by loumalouomega** [comment]
I started with the StructuralMechanicsApplication, but there are many applications that can be migrated too. Right now the current `linear_solver_factory.py` has more options than the `new_linear_solver_factory.py`. It will be nice if people that know how this solvers work can help me with the migration

**Comment by RiccardoRossi** [comment]
Hi @loumalouomega ,
it is not exactly as you say. The point is that the "old" one has all the options in a singe file (the factory itself).

the new one implements the parsing of the parameters in the constructor of each linear solver.

at the very end the options should be the same

**Comment by RiccardoRossi** [comment]
still let's follow @ipouplana's comment and let's ma

**Comment by RiccardoRossi** [comment]
sorry i wanted to delete my comment and closed the issue instead. 

i was telling "let's follow @ipouplana's comment and mention everyone (@KratosMultiphysics) in this issue since it may affect a lot of people

**Comment by pooyan-dadvand** [comment]
I'm ok with this PR. Nevertheless if we want to make a new factory I would suggest to use the factory pattern instead of a monolithic if based file we have now.

**Comment by RiccardoRossi** [comment]
Hi @pooyan-dadvand, in principle i would also agree in the long run to implement a better method for this. The definition of what is a factory pattern sees however to depend a lot on the language and on wether the language implements or not "reflection" (which in this is context is to be intended as the possibility of querying all the subclasses of a given class).

reading into your mind (please correct me if i am wrong) my understanding is that you propose to implement the Factory in c++ rather than in python.

If this is correct, than your proposal would correspond to 
1 -  add a base class LinearSolverFactory with a static singleton "linear_solver_map<name, prototype_obj>" and register all the linear solvers into linear_solver_map.
2 - add to each of the linear solvers a function "MakeSolver" which returns a created object. (also we would need the empty constructor for all the solvers)
3 - define a method MakeSolver of the base class which acts equivalently to the currently existing one.

To my understanding the biggest advantages would be:
 
1. possibility to call the function from within c++ 
2. availability (in c++, in python this could be done already) to get the list of all available solvers.

**Review by rubenzorrilla — COMMENTED** [review]
I'm ok with the enhancements. However, I would take profit of this PR to rename the "new_linear_solver_factory" to "linear_solver_factory" and the old one to "deprecated_linear_solver_factory" to avoid this kind of misunderstandings in the future.

**Comment by pooyan-dadvand** [comment]
As we discussed it just now, the idea is to have a solver factory base class:

```c++
class LinearSolverFactoryBase {
public:
    virtual LinearSolver::Pointer Create(Parameters) = 0;
}
```
and a template factory class which create a solver given as template argument:

```c++
template <typename TLinearSolverType>
class LinearSolverFactory {
public:
    LinearSolver::Pointer Create(Parameters) {
        if(Parameters["Scaling"].GetBool() == true)
            return PointerType(new ScalingSolver(new TLinearSolverType(Parameters)));
        else
            return PointerType(new TLinearSolverType(Parameters));
}
```
and use the factory for registration:
```c++
KRATOS_REGISTER_LINEAR_SOLVER("CGSolver", LinearFactorySolver<CGSolver<Space,...>>)
```

**Comment by RiccardoRossi** [comment]
Just to say that i am convinced that making the registration in python is a good way to go. **I would propose keeping on hold** this pull request until i take a deep look in making it in c++

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi Meanwhile this new factory is an step forward to the old one and keeping this one won't harm anyone until we make it more organic.

**Comment by loumalouomega** [comment]
I will revert this commit in my branch then

**Comment by RiccardoRossi** [comment]
wait, don't do anything yet.

just keep it on hold

**Comment by loumalouomega** [comment]
No, I meant in my branch, that one that is taller and with more commits than the Babylon tower

**Comment by RiccardoRossi** [comment]
with the last commit I did the registration of the linear solvers in c++ as @pooyan-dadvand  suggested

this is **NOT a urgent commit** and should be definitely be postponed to after having finalized the release process (that is, let's leave this for september).

essentially the object linear_solver_factory does not need any longer to be included in python, since it is substituted by 

      solver = KratosMultiphysics.LinearSolverFactoryBase().CreateSolver(configuration)

File: kratos/linear_solvers/linear_solver.h
Comment (RiccardoRossi) [inline kratos/linear_solvers/linear_solver.h:238]: this is defined in all the solvers. I would not change this

**Comment by loumalouomega** [comment]
What do we do with this PR?

**Comment by RiccardoRossi** [comment]
I think that we need some discussion on this. Guys, can you take a look and give feedback?

File: kratos/includes/linear_solver_factory.h
Comment (pooyan-dadvand) [inline kratos/includes/linear_solver_factory.h]: Why these are global functions?

File: kratos/includes/linear_solver_factory.h
Comment (pooyan-dadvand) [inline kratos/includes/linear_solver_factory.h]: And this one?

File: applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp
Comment (pooyan-dadvand) [inline applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp]: We may find a better place to register the linear solvers. i.e. a new `register_linear_solvers.cpp` file

File: kratos/linear_solvers/linear_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/linear_solver.h]: It would be better to move this class to a separate file.

File: kratos/linear_solvers/linear_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/linear_solver.h]: `factory` would be a better name for variable `aux`

File: kratos/python_scripts/new_linear_solver_factory.py
Comment (pooyan-dadvand) [inline kratos/python_scripts/new_linear_solver_factory.py]: Now I'm confused, I think what referred as base is the factory itself and what is referred as factory is a concrete factory.

File: kratos/python_scripts/new_linear_solver_factory.py
Comment (pooyan-dadvand) [inline kratos/python_scripts/new_linear_solver_factory.py]: Lot of ifs removed!! 😄

File: kratos/sources/kernel.cpp
Comment (pooyan-dadvand) [inline kratos/sources/kernel.cpp]: I would strongly avoid calling these functions in the constructor of the Kernel. They are not exception free and with a lot of background details.....

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
In general I liked very much the way @RiccardoRossi implemented the new linear solver factory and now is much more extensible than before. However I would suggest some minor changes to improve it:
- We should find some standard places for registering the linear solvers and preconditioners
- I got confused by the LinearSolverFactoryBase as it seems to be the factory to be used and not the base. 
- I would strongly avoid registering in constructor of Kernel

File: kratos/includes/linear_solver_factory.h
Comment (RiccardoRossi) [reply kratos/includes/linear_solver_factory.h]: where would you put them?

File: kratos/includes/linear_solver_factory.h
Comment (RiccardoRossi) [reply kratos/includes/linear_solver_factory.h]: same, let's agree of where to put them

File: kratos/linear_solvers/linear_solver.h
Comment (RiccardoRossi) [reply kratos/linear_solvers/linear_solver.h]: the point is that this way the class is available wherever you include the linear solver, which was what i wanted to achieve

File: kratos/linear_solvers/preconditioner.h
Comment (RiccardoRossi) [inline kratos/linear_solvers/preconditioner.h]: same argument as for the linear_solver_base. I am putting it in the same archive to have it available where the preconditioenr is available

File: kratos/sources/kernel.cpp
Comment (RiccardoRossi) [reply kratos/sources/kernel.cpp]: I see your point. I think it was needed to put it there, but i don't remember the details. i'll check when i have some time

**Comment by pooyan-dadvand** [comment]
I agree that we need to give it the final iteration after release.

**Comment by pooyan-dadvand** [comment]
Just to add that the overall design is a significant improvment

**Comment by RiccardoRossi** [comment]
@msandre @philbucher @oberbichler here i registered the real-valued solvers but not the complex-valued once, since i was not sure of what should be defined. Could u take a look? it should be trivial to also register the complex valued ones once you know which ones should be registered. 

also note that onces this is done, you could construct the complex linear solver you want directly from c++ within the class FEAST solver simply passing the pertinent parameters

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi I think we should create a new `Registry` class to centralize all registrations in Kratos. Actually this message is reminder note for myself! :smile:

**Comment by oberbichler** [comment]
Who is using the complex solvers?

Is it possible to register the solvers from within the application and without modifying the …factory.py?

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand let's discuss your proposal on monday. Also to be honest i don't understand how you would like the current registration mechanism to be changed, so i did not address some of your suggestions

@oberbichler yes, the idea is to register the solvers at the c++ level, so that the _factory.py becomes trivial. The main advantages are that registration is done at compile time and that the list of available solvers is automatically updated. Also it is useful to have the possibility of using the factory from c++

**Comment by philbucher** [comment]
@qaumann is using complex solvers afaik

**Comment by qaumann** [comment]
@philbucher The FEAST solver needs a complex linear solver (see fest_solver.h). I am not using a complex solver outside of FEAST

**Comment by msandre** [comment]
There are two complex solvers for FEAST:

- A default one which uses the skyline from amgcl. 

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_linear_solvers_to_python.cpp#L212

- a faster one that depends on the tpl pastix:

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp#L102

we need to add both to the registry. It should simply be a matter of registering the solvers and running the eigen tests in structural mechanics application to see that it works. I can't do it today but will try to do it soon.

**Comment by msandre** [comment]
The complex solvers are added now. @philbucher the check you added to Eigen_Q4_Thick_2x2_Plate_test_parameters.json fails on my machine. It looks like the numbers are affected by roundoff. Can you look at it?

**Comment by msandre** [comment]
Also @RiccardoRossi can you check my changes to the scaling solver? I had to modify it so it would compile with the complex solvers.

**Comment by philbucher** [comment]
@msandre I know that it fails, fixing it is on my todo list

**Comment by RiccardoRossi** [comment]
ping @pooyan-dadvand

File: applications/ExternalSolversApplication/external_includes/feast_solver.h
Comment (RiccardoRossi) [inline applications/ExternalSolversApplication/external_includes/feast_solver.h:105]: i review myself since a lot of time passed: should this be complex_skyline_lu_solver?

File: kratos/sources/linear_solver_factory.cpp
Comment (RiccardoRossi) [inline kratos/sources/linear_solver_factory.cpp]: @pooyan-dadvand ,maybe we should call this 

       SkylineLUComplexSolverFunctor

instead of 

       SkylineLUComplexSolverFactory

File: kratos/sources/variables.cpp
Comment (RiccardoRossi) [inline kratos/sources/variables.cpp]: i am not sure if this is the correct place, however we need somewher that is surely called at initialization time

**Review by RiccardoRossi — COMMENTED** [review]
@pooyan-dadvand i read the book you sent me and i would say that the implementation is ok.

most probably we could rename some stuff to "Functor" and call the FactoryBase as "AbstractFactory" but other than that i think that the implementation is more or less ok.

**Comment by pooyan-dadvand** [comment]
let me check it again, I forgot my comments on this

**Comment by loumalouomega** [comment]
Which is the status of this?

**Comment by pooyan-dadvand** [comment]
I should give it the last polish with @RiccardoRossi but after the release

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand this is getting stale. It would be nice if this could be in for release 6...

**Comment by pooyan-dadvand** [comment]
I agree with you. The delay is my fault. Let see it together next week and finish it.

**Comment by roigcarlo** [comment]
Hi, this is flagged for 6.0 but I see a lot of conflicts here. I will remove this from the release. If this is ready for 6.0 please re add it next week and we will make a separate Pr into the fixed branch

File: applications/ExternalSolversApplication/CMakeLists.txt
Comment (pooyan-dadvand) [inline applications/ExternalSolversApplication/CMakeLists.txt]: I would put file in the linear_solvers folder. Meanwhile it is almost empty...

File: applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp
Comment (pooyan-dadvand) [inline applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp]: At this stage you don't need the `RegisterLinearSolvers `at all. You may do it in application registration until I implement the Registry class in the core.

File: applications/ExternalSolversApplication/custom_utilities/register_linear_solvers.h
Comment (pooyan-dadvand) [inline applications/ExternalSolversApplication/custom_utilities/register_linear_solvers.h]: You don't need this class as long as using `KratosComponents `as your registry.

File: applications/ExternalSolversApplication/externalsolvers_application.cpp
Comment (pooyan-dadvand) [inline applications/ExternalSolversApplication/externalsolvers_application.cpp]: At this stage you may put your the class constructor directly here.

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: It is not a substitution of the old one? without new prefix I mean.

File: kratos/linear_solvers/bicgstab_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/bicgstab_solver.h]: I would make the `CreatePreconditioner `a method of the `BaseType` (even with the check `Has("preconditioner_type")` included.

Comment (pooyan-dadvand) [reply kratos/linear_solvers/bicgstab_solver.h]: And here I would call the CreatePreconditioner in the base class constructor to avoid code duplication.

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Now I understood the structure and I think that the confusing part (for me) was the fact that in your implementation the ObjectFactory is the LinearSolverFactoryBase and the ObjectCreator is the LinearSolverFactory derived from the base one. Meanwhile to have the work reusable after the implementation of Registery and ObjectFactory, it would be better to have Create(Parameters) methods for each solver. Nevertheless I understand if you decide to postponed this change to when we have the registery.

They are some minor comments also in the code.

**Comment by loumalouomega** [comment]
I created this PR, but I think that most of that changes are from @RiccardoRossi , should I do something?

**Comment by pooyan-dadvand** [comment]
My comments are for @RiccardoRossi

**Comment by loumalouomega** [comment]
I solved the conflicts to tackle this again, I hope I haven't broke anything.

**Comment by loumalouomega** [comment]
Some leftovers of boost python are needed to be cleaned

**Comment by loumalouomega** [comment]
@RiccardoRossi and @pooyan-dadvand I fixed the compilation and now all the tests work. I added the EigenApplicationSolvers too

**Comment by loumalouomega** [comment]
Ok, right now the only remaining issue is the cotire compilation, @RiccardoRossi can you check that?

**Comment by philbucher** [comment]
you sure its cotire?
You are missing some overrides which are turned into errors

**Comment by loumalouomega** [comment]
that explains a lot :), I was reading cotire messages, sorry for the noise

**Comment by philbucher** [comment]
when Travis doesn't compile take a close look at the output => since #2572  some warnings are turned into errors in order to keep the code clean

**Comment by loumalouomega** [comment]
OK

**Comment by loumalouomega** [comment]
This is ready for merge

**Comment by oberbichler** [comment]
👍 Looks good!

About the complex solvers in the `EigenSolverApplication`: I think they are testest (@armingeiser ?)

Maybe we should drop the old solver names.

**Comment by oberbichler** [comment]
...and why `std::string("...")` instead of `"..."`?

**Comment by loumalouomega** [comment]
Where is that std::string?

File: applications/EigenSolversApplication/custom_python/add_custom_solvers_to_python.cpp
Comment (oberbichler) [inline applications/EigenSolversApplication/custom_python/add_custom_solvers_to_python.cpp]: here

**Comment by armingeiser** [comment]
@oberbichler The complex solvers from the `EigenSolversApplication` are only tested within FEAST (the same way as also ComplexSkylineLU).

**Comment by oberbichler** [comment]
👍  I was just wondering why `ComplexLDLT/LLT` are commented out but they are not covered by the FEAST test...

**Comment by loumalouomega** [comment]
Done

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: actually here i would call directly

       KratosMultiphysics.LinearSolverFactoryBase().CreateSolver(configuration)

**Review by RiccardoRossi — COMMENTED** [review]
@loumalouomega to me this looks ok. Thank you very much for taking the time to fix this.

@pooyan-dadvand it is on you to approve this ...

**Comment by loumalouomega** [comment]
I solved the conflict with AMatrix, maybe we should merge this before more conflicts arrise

**Comment by loumalouomega** [comment]
@pooyan-dadvand this is ready

**Comment by loumalouomega** [comment]
Do I apply the changes related with AMatrix to this branch or do I create a new PR after this is merged?

**Comment by loumalouomega** [comment]
Conflict solved @pooyan-dadvand

**Comment by loumalouomega** [comment]
@pooyan-dadvand I would like to have this merged in order to develop a proper adaptive NR. I discussed with @RiccardoRossi and the idea is to extend the idea of the factories to all the components integrating the NR (B&S, convergence criterion, etc...). That way we could create a constructor for the NR based on parameters instead of adding more arguments as doing right now.

**Comment by pooyan-dadvand** [comment]
@loumalouomega we can check this together in Wednesday when I pass by CIMNE

**Comment by loumalouomega** [comment]
Ok :+1:

File: kratos/includes/linear_solver_factory.h
Comment (pooyan-dadvand) [inline kratos/includes/linear_solver_factory.h:63]: `LinearSolverFactory` to be changed to `StandardLinearSolverFactory`

File: kratos/includes/linear_solver_factory.h
Comment (pooyan-dadvand) [inline kratos/includes/linear_solver_factory.h]: `CreateHelper` -> `CreateSolver`

File: kratos/includes/linear_solver_factory.h
Comment (pooyan-dadvand) [reply kratos/includes/linear_solver_factory.h]: Each registration in the corresponding header/cpp file

File: kratos/linear_solvers/linear_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/linear_solver.h]: `LinearSolverFactoryBase` to `LinearSolverFactory`

File: kratos/linear_solvers/linear_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/linear_solver.h]: `CreateSolver` to `Create`

File: kratos/linear_solvers/linear_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/linear_solver.h]: `CreateHelper` to `CreateSolver`

File: kratos/linear_solvers/preconditioner.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/preconditioner.h]: Same for this base

File: kratos/linear_solvers/scaling_solver.h
Comment (pooyan-dadvand) [inline kratos/linear_solvers/scaling_solver.h]: take out the comments please

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I would add macro for registering the solvers and preconditioners and distribute them in their corresponding files

**Comment by loumalouomega** [comment]
Some changes are still pendent. I will upload them later

**Comment by loumalouomega** [comment]
The only thing I am not able yet is to register in each separated file without conflict python register

**Comment by loumalouomega** [comment]
@pooyan-dadvand @roigcarlo told me that is not possible to directly register the solvers in the header, it is necessary at least to call a cpp

**Review by pooyan-dadvand — APPROVED** [review]
So we postpone the distributed registration to when Kratos Factory is implemented

**Comment by loumalouomega** [comment]
Hurray

File: kratos/solving_strategies/strategies/solving_strategy.h
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h:351]: why did you remove this?

**Review by philbucher — COMMENTED** [review]
Finally :D

You have not updated / registered te trilinos solvers right?

**Comment by loumalouomega** [comment]
I think the Trilinos are not registered, I have not register them at least

**Comment by pooyan-dadvand** [comment]
Yes, this one become even older than me! :stuck_out_tongue_winking_eye:

**Comment by philbucher** [comment]
one question guys:
to use it for the mpi-solvers, one should derive a `DistributedLinearSolverFactory` from the `LinearSolverFactory` right?
same as what is done for the `StandardLinearSolverFactory` (weird name btw)

**Comment by loumalouomega** [comment]
Yes, exactly, yesterday I tried to compile Trilinos and fix it, but I was unable. Are you suing Ubuntu?

**Comment by pooyan-dadvand** [comment]
> same as what is done for the StandardLinearSolverFactory (weird name btw)

Any suggestion?

**Comment by loumalouomega** [comment]
I will mention #3185 as this PR moves the factories to one folder, and can be an opportunity to change the name of the standard factory

**Comment by loumalouomega** [comment]
BTW linear solver factories are added to Trilinos in here:

#3204

**Pull request merged**
